### ISSUE TITLE
Umbraco backoffice node_modules content exceeds OS path limit

### ISSUE THREAD

**Issue by gregor-tusar-sowa** [umbraco/Umbraco-CMS#16659]
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

14.0.0

### Bug summary

While updating custom dashboard from v13 to v14 we stumbled upon an issue that is caused by Umbraco `@umbraco-cms/backoffice` package.

While trying to build the solution a lot of paths exceed the OS limit of 260.

Shortening some of our names did not help, because after that copying to output directory fails.

The paths of some of the `node_modules` files itself is 185 char long, so adding drive letter, some folders to solution and build output folder `bin\Debug\net8.0\` quickly exceeds OS max limit.

### Steps to reproduce

- Create a solution in path `C:\src\ORG\MyComp\src\Cms\Company.Cms`.  
- Create a [new dashboard](https://docs.umbraco.com/umbraco-cms/tutorials/creating-a-custom-dashboard) in `App_Plugins` with [the new approach](https://docs.umbraco.com/umbraco-cms/extending/customize-backoffice/vite-package-setup) 
- Try to build

### Expected result / actual result

We expect the solution would successfully build.

ACTUAL RESULT:
```
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-grid\property-editors\block-grid-type-configuration\property-editor-ui-block-grid-type-configuration.element.js" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-grid\property-editors\block-grid-type-configuration\property-editor-ui-block-grid-type-configuration.element.js". Exceeded retry count of 10. Failed. 	
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-grid\property-editors\block-grid-type-configuration\property-editor-ui-block-grid-type-configuration.element.d.ts" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-grid\property-editors\block-grid-type-configuration\property-editor-ui-block-grid-type-configuration.element.d.ts". Exceeded retry count of 10. Failed. 	
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-grid\property-editors\block-grid-group-configuration\property-editor-ui-block-grid-group-configuration.element.js" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-grid\property-editors\block-grid-group-configuration\property-editor-ui-block-grid-group-configuration.element.js". Exceeded retry count of 10. Failed. 	
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-grid\property-editors\block-grid-group-configuration\property-editor-ui-block-grid-group-configuration.element.d.ts" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-grid\property-editors\block-grid-group-configuration\property-editor-ui-block-grid-group-configuration.element.d.ts". Exceeded retry count of 10. Failed. 	
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-grid\property-editors\block-grid-layout-stylesheet\property-editor-ui-block-grid-layout-stylesheet.element.d.ts" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-grid\property-editors\block-grid-layout-stylesheet\property-editor-ui-block-grid-layout-stylesheet.element.d.ts". Exceeded retry count of 10. Failed. 	
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-grid\property-editors\block-grid-layout-stylesheet\property-editor-ui-block-grid-layout-stylesheet.element.js" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-grid\property-editors\block-grid-layout-stylesheet\property-editor-ui-block-grid-layout-stylesheet.element.js". Exceeded retry count of 10. Failed. 	
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-list\property-editors\block-list-type-configuration\property-editor-ui-block-list-type-configuration.element.js" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-list\property-editors\block-list-type-configuration\property-editor-ui-block-list-type-configuration.element.js". Exceeded retry count of 10. Failed. 	
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-list\property-editors\block-list-type-configuration\property-editor-ui-block-list-type-configuration.element.d.ts" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-list\property-editors\block-list-type-configuration\property-editor-ui-block-list-type-configuration.element.d.ts". Exceeded retry count of 10. Failed. 	
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-type\property-editors\block-type-group-configuration\property-editor-ui-block-type-group-configuration.element.d.ts" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-type\property-editors\block-type-group-configuration\property-editor-ui-block-type-group-configuration.element.d.ts". Exceeded retry count of 10. Failed. 	
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-type\property-editors\block-type-group-configuration\property-editor-ui-block-type-group-configuration.element.js" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\block\block-type\property-editors\block-type-group-configuration\property-editor-ui-block-type-group-configuration.element.js". Exceeded retry count of 10. Failed. 	
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\core\workspace\components\workspace-breadcrumb\workspace-variant-menu-breadcrumb\workspace-variant-menu-breadcrumb.element.js" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\core\workspace\components\workspace-breadcrumb\workspace-variant-menu-breadcrumb\workspace-variant-menu-breadcrumb.element.js". Exceeded retry count of 10. Failed. 	
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\core\workspace\components\workspace-breadcrumb\workspace-variant-menu-breadcrumb\workspace-variant-menu-breadcrumb.element.d.ts" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\core\workspace\components\workspace-breadcrumb\workspace-variant-menu-breadcrumb\workspace-variant-menu-breadcrumb.element.d.ts". Exceeded retry count of 10. Failed. 	
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\core\workspace\components\workspace-breadcrumb\workspace-variant-menu-breadcrumb\workspace-variant-menu-breadcrumb.kind.js" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\core\workspace\components\workspace-breadcrumb\workspace-variant-menu-breadcrumb\workspace-variant-menu-breadcrumb.kind.js". Exceeded retry count of 10. Failed. 	
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\core\workspace\components\workspace-breadcrumb\workspace-variant-menu-breadcrumb\workspace-variant-menu-breadcrumb.kind.d.ts" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\core\workspace\components\workspace-breadcrumb\workspace-variant-menu-breadcrumb\workspace-variant-menu-breadcrumb.kind.d.ts". Exceeded retry count of 10. Failed. 	
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\documents\documents\user-permissions\input-document-granular-user-permission\input-document-granular-user-permission.element.d.ts" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\documents\documents\user-permissions\input-document-granular-user-permission\input-document-granular-user-permission.element.d.ts". Exceeded retry count of 10. Failed. 	
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\documents\documents\user-permissions\input-document-granular-user-permission\input-document-granular-user-permission.element.js" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\documents\documents\user-permissions\input-document-granular-user-permission\input-document-granular-user-permission.element.js". Exceeded retry count of 10. Failed. 	
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\property-editors\content-picker\dynamic-root\input-content-picker-document-root\input-content-picker-document-root.element.d.ts" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\property-editors\content-picker\dynamic-root\input-content-picker-document-root\input-content-picker-document-root.element.d.ts". Exceeded retry count of 10. Failed. 	
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\property-editors\content-picker\dynamic-root\input-content-picker-document-root\input-content-picker-document-root.element.js" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\property-editors\content-picker\dynamic-root\input-content-picker-document-root\input-content-picker-document-root.element.js". Exceeded retry count of 10. Failed. 	
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\templating\modals\templating-page-field-builder\components\template-field-dropdown-list\template-field-dropdown-list.element.js" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\templating\modals\templating-page-field-builder\components\template-field-dropdown-list\template-field-dropdown-list.element.js". Exceeded retry count of 10. Failed. 	
Error (active)	MSB3027	Could not copy "C:\src\ORG\MyComp\src\Cms\Company.Cms\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\templating\modals\templating-page-field-builder\components\template-field-dropdown-list\template-field-dropdown-list.element.d.ts" to "bin\Debug\net8.0\App_Plugins\welcome-dashboard\node_modules\@umbraco-cms\backoffice\dist-cms\packages\templating\modals\templating-page-field-builder\components\template-field-dropdown-list\template-field-dropdown-list.element.d.ts". Exceeded retry count of 10. Failed.
```

**Comment by iOvergaard** [comment]
Hello, and thanks for reporting this.

We are planning to shorten the paths, but it is challenging to get under the abstract path length of 260 characters in Windows. Up until now, we have only seen the problem with regards to the Backoffice files in the Umbraco StaticAssets RCL. What I do not fully grasp here is why Visual Studio (I presume) is copying "node_modules" to your "bin/debug" folder. That is certainly not needed to build an extension for Umbraco.

There is a similar situation described here, [where an Angular project exceeded the file length of Windows](https://stackoverflow.com/a/55647048) and a proposed solution. I think that could work for your situation, would you be willing to try it out?

**Comment by gregor-tusar-sowa** [comment]
I managed to solve this issue with updating the MSBuild version to 17.10.
The issue was that Visual Studio took MSBuild version 16, which failed to copy long file paths.
After changing the path to the MSBuild in environmental variables to the newest MSBuild the build went through smoothly.

**Comment by iOvergaard** [comment]
Oh, good to know @gregor-tusar-sowa! 
Just be aware that you don't need to have node_modules in your output folder for this to work, so you can save a few bytes by excluding it.

**Comment by gregor-tusar-sowa** [comment]
Thanks for letting me know that.

Adding this `ItemGroup` to my project does not exclude the `node_modules` folder to being copied to output and I don't know why.

```
  <ItemGroup>
    <Content Remove="App_Plugins\**\node_modules\**" />
    <Compile Remove="App_Plugins\**\node_modules\**" />
    <EmbeddedResource Remove="App_Plugins\**\node_modules\**" />
    <None Remove="App_Plugins\**\node_modules\**" />
  </ItemGroup>
  ```

**Comment by iOvergaard** [comment]
@gregor-tusar-sowa I am still looking into this to see, how you could exclude `node_modules` and other unwanted files. It looks like the Umbraco.Cms.Targets overrule whatever you try to do since it will copy App_Plugins blindly following this:

https://github.com/umbraco/Umbraco-CMS/blob/3ce2e97f5e13de0f4e69ff8943c66744d3103b27/src/Umbraco.Cms.Targets/buildTransitive/Umbraco.Cms.Targets.targets#L32-L38

I recommend putting the `node_modules` and other source files outside of App_Plugins, perhaps in a Client folder, or even better in an RCL, where you get a `wwwroot` folder, which you can copy the compiled files to.

I realise that our documentation does not state this, so that is something we are going to look into getting added!

**Comment by gregor-tusar-sowa** [comment]
Thanks for looking into this. We will try to move things around a bit.

**Comment by wadrschall** [comment]
Is there any update on this issue? We are trying to upgrade a client from 13 to 14 and are having similar issues:

`Could not copy "C:\Users\builduser\.nuget\packages\umbraco.cms.staticassets\14.1.1\staticwebassets\umbraco\backoffice\packages\property-editors\content-picker\dynamic-root\input-content-picker-document-root\input-content-picker-document-root.element.d.ts" to "C:\AzureDevOps\agents\Self Host\BS-AGT-4\_work\1013\s\FolderName\obj\Release\net8.0\PubTmp\Out\wwwroot\umbraco\backoffice\packages\property-editors\content-picker\dynamic-root\input-content-picker-document-root\input-content-picker-document-root.element.d.ts". Exceeded retry count of 10. Failed. `

We run our builds via DevOps on our own internal dev servers (windows based) and have this issue. We have no issues with any of our 13 sites. 

Please advise the planned fix/workaround for this issue. I assume the plan isn't to force all developers to build using linux based build servers in order to handle umbraco's new extra long file path names.

**Comment by iOvergaard** [comment]
@wadrschall One solution is to use a minimum of MSBuild 17, which supports longer paths. 
The upcoming version 14.2 will also bring the length of our files down to a max of 120 characters leaving more room for the rest of the operating system path. This will also better support more build tools. 

Eventually, when we have managed to bundle all of the backoffice modules, we will have more or less a flat structure, which I believe will be the end of this problem. 

I hope everyone can appreciate that there is no golden rule as to how long the paths of a .NET library can or should be.

**Comment by nul800sebastiaan** [comment]
Just to follow-up, this specific problem should not exists and I've attempted a fix in #17331 

I know we've documented this setup of putting all your plugins in `App_Plugins`, and that needs to be updated as well since it will also copy all the source files (including `node_modules)` into the bin folder and publish it all to a server (!). This is definitely not what we should want, that's like publishing all the nuget files during a deploy.

So I'll look at getting the docs updated as well here, since this all can be avoided.
